grpc/tests/src/test/gas_counter.rs

```diff
@@ -0,0 +1,236 @@
+            instructions.push(Instruction::Nop);
+        }
+        instructions.push(Instruction::End);
+        Instructions::new(instructions)
+    };
+
+    let module = builder::module()
+        .function()
+        // A signature with 0 params and no return type
+        .signature()
+        .build()
+        .body()
+        // Generated instructions for our entrypoint
+        .with_instructions(instructions)
+        .build()
+        .build()
+        // Export above function
+        .export()
+        .field(DEFAULT_ENTRY_POINT_NAME)
+        .build()
+        // Memory section is mandatory
+        .memory()
+        .build()
+        .build();
+    parity_wasm::serialize(module).expect("should serialize")
+}
+
+/// Creates session code with opcodes
+fn make_session_code_with(instructions: Vec<Instruction>) -> Vec<u8> {
+    let module = builder::module()
+        .function()
+        // A signature with 0 params and no return type
+        .signature()
+        .build()
+        .body()
+        // Generated instructions for our entrypoint
+        .with_instructions(Instructions::new(instructions))
+        .build()
+        .build()
+        // Export above function
+        .export()
+        .field(DEFAULT_ENTRY_POINT_NAME)
+        .build()
+        // Memory section is mandatory
+        .memory()
+        .build()
+        .build();
+    parity_wasm::serialize(module).expect("should serialize")
+}
+
+#[ignore]
+#[test]
+fn should_fail_to_overflow_gas_counter() {
+    let mut builder = InMemoryWasmTestBuilder::default();
+
+    let session_bytes = make_gas_counter_overflow();
+
+    let exec_request = {
+        let deploy_item = DeployItemBuilder::new()
+            .with_address(*DEFAULT_ACCOUNT_ADDR)
+            .with_session_bytes(session_bytes, RuntimeArgs::default())
+            .with_empty_payment_bytes(runtime_args! {
+                ARG_AMOUNT => *DEFAULT_PAYMENT
+            })
+            .with_authorization_keys(&[*DEFAULT_ACCOUNT_ADDR])
+            .with_deploy_hash([42; 32])
+            .build();
+        ExecuteRequestBuilder::from_deploy_item(deploy_item).build()
+    };
+
+    builder.run_genesis(&DEFAULT_RUN_GENESIS_REQUEST);
+
+    builder.exec(exec_request).commit();
+
+    let responses = builder.get_exec_response(0).expect("should have response");
+    let response = responses.get(0).expect("should have first element");
+
+    let lhs = response.as_error().expect("should have error");
+    assert_matches!(
+        lhs,
+        Error::WasmPreprocessing(PreprocessingError::OperationForbiddenByGasRules)
+    );
+}
+
+#[ignore]
+#[test]
+fn should_correctly_measure_gas_for_opcodes() {
+    let opcode_costs = DEFAULT_WASM_CONFIG.opcode_costs();
+
+    const GROW_PAGES: u32 = 1;
+
+    let opcodes = vec![
+        (Some(opcode_costs.nop), Instruction::Nop),
+        (
+            Some(opcode_costs.current_memory),
+            Instruction::CurrentMemory(0),
+        ), // Push size to stack
```

Comment:
I feel this test would benefit from some more comments explaining the rationale of the test itself, and in particular what this Vec represents.

Reply:
Did this